### ISSUE TITLE
Namespace not found issues

### ISSUE THREAD

**Issue by TheAngryByrd** [ionide/ionide-fsharp#146]
I tried opening one of my projects but I end up with a lot of namespace not found errors.

![screen shot 2015-09-29 at 4 22 42 pm](https://cloud.githubusercontent.com/assets/1490044/10177110/7a45b22e-66c6-11e5-8320-e8b6fe39fad9.png)
![screen shot 2015-09-29 at 4 22 53 pm](https://cloud.githubusercontent.com/assets/1490044/10177111/7a4c4e22-66c6-11e5-9eb3-e0cc4dbea842.png)

Xamarin Studio and Visual Studio have no complaints.  Am I missing something obvious? 

Thanks!

**Comment by Krzysztof-Cieslak** [comment]
That's interesting. Looks like a problem with FSAC? /cc: @rneatherway

**Comment by rneatherway** [comment]
Is it possible to paste the output of the `project` command to FSAC? And also the fsproj file would be useful.

**Comment by TheAngryByrd** [comment]
I'll come up with a sample project and get back to you.

**Comment by TheAngryByrd** [comment]
I created the project like I did with my other projects.  I used ProjectScaffold template, added dependencies with paket on command line, and edited some code with Xamarin Studio, then opened with Atom.

https://github.com/TheAngryByrd/IonideBugSample

**Comment by rneatherway** [comment]
What OS are you running. If Windows, which versions of VS are installed. If not, which versions of Mono and F#?

**Comment by TheAngryByrd** [comment]
Yeah, forgot to mention, OSX.

**Comment by TheAngryByrd** [comment]
I do have the 32 bit version of Mono because Xamarin Studio packages that.

**Comment by rneatherway** [comment]
I see, could you also please verify your version of fsautocomplete?

**Comment by TheAngryByrd** [comment]
FsAutoComplete 0.24.1

**Comment by tonyj-dootrix** [comment]
I think I'm experiencing the same issue, using the fsharp yeoman generator to create a Websharper.UI.Next project.  When I open Client.fs I get namespace not found errors for WebSharper (Paket dependency - added and installed by the yo generator).

![image](https://cloud.githubusercontent.com/assets/14938281/11641611/16fa430a-9d31-11e5-8213-0df8b56ab347.png)

I'm also using 32-bit Mono 4.03 on OSX.

**Comment by Krzysztof-Cieslak** [comment]
I can reproduce it on Windows - it's working in VS, while Ionide (both Atom and VS Code) shows me same errors. What makes me believe it's some problem with FSAC.

**Comment by rneatherway** [comment]
WebSharper has caused some problems in the past because it uses a custom MSBuild task. This might be fixed by upgrading FSAC to FCS 2.0.0, where I reworked the project cracking. I have a branch nearly finished for that and will do a new release soon.

**Comment by TheAngryByrd** [comment]
I'll definitely be ready for testing it.

**Comment by rneatherway** [comment]
I have released a new version of FSAC now, but it probably isn't included in ionide yet. You could get some debug output by downloading [the release](https://github.com/fsharp/FsAutoComplete/releases/download/0.27.0/fsautocomplete.zip), unzipping it and running something like:

    FSharp.Compiler.Service.ProjectCrackerTool.exe --text <path to>/MyProject.fsproj true

**Comment by Krzysztof-Cieslak** [comment]
Actually, I've already updated to `FSAC 0.27.0`

**Comment by rneatherway** [comment]
Cool! In that case to get debug output the project command needs to be told to be verbose, have you added support for that?

**Comment by Krzysztof-Cieslak** [comment]
No, not yet. I've only updated FSAC haven't added new functionalities yet.

**Comment by rneatherway** [comment]
Ok no worries. @TheAngryByrd @tonyj-dootrix  if you have a chance to do the manual test I suggested above that would be great.

**Comment by tonyj-dootrix** [comment]
[debug.txt](https://github.com/ionide/ionide-fsharp/files/60935/debug.txt)


This help?

**Comment by TheAngryByrd** [comment]
Ok I'll give it a shot.

**Comment by rneatherway** [comment]
@tonyj-dootrix there don't seem to be any errors and the WebSharper DLLs are found, so I think it will be working with the latest version.

**Comment by tonyj-dootrix** [comment]
Have updated ionide-fsharp to 1.5.0 but I'm still getting the same errors in Atom.  Is there something else I should be updating?

**Comment by rneatherway** [comment]
@TheAngryByrd I tried your example project and it works fine for me on Linux Mono 4.2.

@tonyj-dootrix can you make your project available for me to double-check please?

**Comment by tonyj-dootrix** [comment]
[wsdemo.zip](https://github.com/ionide/ionide-fsharp/files/61436/wsdemo.zip)

This is a completely 'clean' project, simply generated using the f-sharp yeoman generator.  I chose the 'Create standalone project' option and then chose 'WebSharper.UI.Next SPA'.  Let  me know if there's anything else you need.

**Comment by rneatherway** [comment]
Here are the dependencies I find locally:

```
    "-r:packages/WebSharper/lib/net40/IntelliFactory.Xml.dll";
    "-r:packages/WebSharper/lib/net40/WebSharper.Collections.dll";
    "-r:packages/WebSharper/lib/net40/WebSharper.Control.dll";
    "-r:packages/WebSharper/lib/net40/WebSharper.Core.JavaScript.dll";
    "-r:packages/WebSharper/lib/net40/WebSharper.Core.dll";
    "-r:packages/WebSharper/lib/net40/WebSharper.Html.Client.dll";
    "-r:packages/WebSharper/lib/net40/WebSharper.Html.Server.dll";
    "-r:packages/WebSharper/lib/net40/WebSharper.InterfaceGenerator.dll";
    "-r:packages/WebSharper/lib/net40/WebSharper.JQuery.dll";
    "-r:packages/WebSharper/lib/net40/WebSharper.JavaScript.dll";
    "-r:packages/WebSharper/lib/net40/WebSharper.Main.dll";
    "-r:packages/WebSharper/lib/net40/WebSharper.Sitelets.dll";
    "-r:packages/WebSharper/lib/net40/WebSharper.Testing.dll";
    "-r:packages/WebSharper/lib/net40/WebSharper.Web.dll";
    "-r:packages/WebSharper.UI.Next/lib/net40/WebSharper.UI.Next.Templating.dll";
    "-r:packages/WebSharper.UI.Next/lib/net40/WebSharper.UI.Next.dll";
    "-r:/usr/lib/mono/4.5/mscorlib.dll";
    "-r:/usr/lib/mono/gac/FSharp.Core/4.3.1.0__b03f5f7f11d50a3a/FSharp.Core.dll";
    "-r:/usr/lib/mono/4.5/System.dll";
    "-r:/usr/lib/mono/4.5/System.Numerics.dll";
    "-r:/usr/lib/mono/4.5/System.Web.dll"; "-r:/usr/lib/mono/4.5/System.Xml.dll";
    "-r:/usr/lib/mono/4.5/System.Xml.Linq.dll";
    "-r:/usr/lib/mono/4.5/System.Core.dll";
```

And it works fine in the latest version of Emacs, so I think it is an issue with the Suave frontend. I'm not sure why there would be a problem though.

**Comment by tonyj-dootrix** [comment]
Sorry, I don't understand how Suave is involved?

**Comment by rneatherway** [comment]
Ionide uses a Suave interface to FsAutoComplete rather than a pipe like Emacs and Vim. @Krzysztof-Cieslak will know how to debug that properly.

**Comment by Krzysztof-Cieslak** [comment]
@rneatherway 

I can reproduce (at least on some of my projects) it on my PC. After update to newest FSAC (so with new resolver).

For example:
https://github.com/deepcovelabs/ServiceFabricSuave/blob/master/src/CalculatorService/CalculatorService.fsproj
gives:
```
{"Kind":"project","Data":{"Project":"d:\\Programowanie\\!DCL\\HelloWorld\\src\\CalculatorService\\CalculatorService.fsproj","Files":["d:\\Programowanie\\!DCL\\HelloWorld\\src\\CalculatorService\\AssemblyInfo.fs","d:\\Programowanie\\!DCL\\HelloWorld\\src\\CalculatorService\\ServicesCommon.fs","d:\\Programowanie\\!DCL\\HelloWorld\\src\\CalculatorService\\Services.fs","d:\\Programowanie\\!DCL\\HelloWorld\\src\\CalculatorService\\Program.fs"],"Output":"CalculatorService.exe","References":["C:\\Program Files (x86)\\Reference Assemblies\\Microsoft\\FSharp\\.NETFramework\\v4.0\\4.4.0.0\\FSharp.Core.dll"]}}
```
which obviously has some more references than just FSharp.Core ;)

**Comment by rneatherway** [comment]
Please could you run it again in "verbose" mode? Just add an extra 'verbose' to the end of the 'project' command, or run it manually as per https://github.com/fsharp/emacs-fsharp-mode#project-file-issues

**Comment by Krzysztof-Cieslak** [comment]
Output running verbose option:
```
project "d:\\Programowanie\\!DCL\\HelloWorld\\src\\CalculatorService\\CalculatorService.fsproj" verbose
{"Kind":"project","Data":{"Project":"d:\\Programowanie\\!DCL\\HelloWorld\\src\\CalculatorService\\CalculatorService.fsproj","Files":["d:\\Programowanie\\!DCL\\HelloWorld\\src\\CalculatorService\\AssemblyInfo.fs","d:\\Programowanie\\!DCL\\HelloWorld\\src\\CalculatorService\\ServicesCommon.fs","d:\\Programowanie\\!DCL\\HelloWorld\\src\\CalculatorService\\Services.fs","d:\\Programowanie\\!DCL\\HelloWorld\\src\\CalculatorService\\Program.fs"],"Output":"CalculatorService.exe","References":["C:\\Program Files (x86)\\Reference Assemblies\\Microsoft\\FSharp\\.NETFramework\\v4.0\\4.4.0.0\\FSharp.Core.dll"],"Logs":{"d:\\Programowanie\\!DCL\\HelloWorld\\src\\CalculatorService\\CalculatorService.fsproj":"Build started.\r\nProject \"CalculatorService.fsproj\" (Build target(s)):\r\nBuilding with tools version \"12.0\".\r\nProject file contains ToolsVersion=\"14.0\". This toolset may be unknown or missing, in which case you may be able to resolve this by installing the appropriate version of MSBuild, or the build may have been forced to a particular ToolsVersion for policy reasons. Treating the project as if it had ToolsVersion=\"12.0\". For more information, please see http://go.microsoft.com/fwlink/?LinkId=293424.\r\nTarget \"_CheckForInvalidConfigurationAndPlatform\" in file \"C:\\Program Files (x86)\\MSBuild\\12.0\\bin\\Microsoft.Common.CurrentVersion.targets\" from project \"d:\\Programowanie\\!DCL\\HelloWorld\\src\\CalculatorService\\CalculatorService.fsproj\" (entry point):\r\nUsing \"Error\" task from assembly \"Microsoft.Build.Tasks.v12.0, Version=12.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a\".\r\nTask \"Error\"\r\nThe OutputPath property is not set for project 'CalculatorService.fsproj'.  Please check to make sure that you have specified a valid combination of Configuration and Platform for this project.  Configuration='Debug'  Platform='AnyCPU'.
  This error may also appear if some other project is trying to follow a project-to-project reference to this project, this project has been unloaded or is not included in the solution, and the referencing project does not build using the same or an equivalent Configuration or Platform.\r\nDone executing task \"Error\" -- FAILED.\r\nDone building target \"_CheckForInvalidConfigurationAndPlatform\" in project \"CalculatorService.fsproj\" -- FAILED.\r\nDone building project \"CalculatorService.fsproj\" -- FAILED.\r\nBuild FAILED.\r\n"}}}
```

**Comment by rneatherway** [comment]
What platform? If windows, which versions of VS and MSBuild are installed? Is the OutputPath property specified in the project file?

**Comment by Krzysztof-Cieslak** [comment]
Windows 10, VS 2015, MSBuild 14 and 12. Output path is set [here](https://github.com/deepcovelabs/ServiceFabricSuave/blob/master/src/CalculatorService/CalculatorService.fsproj#L25) and [here](https://github.com/deepcovelabs/ServiceFabricSuave/blob/master/src/CalculatorService/CalculatorService.fsproj#L38)

**Comment by Krzysztof-Cieslak** [comment]
Please also consider that it's output from command line version of FSAC so it's not Suave version problem

**Comment by Krzysztof-Cieslak** [comment]
(Also even if something would be wrong with output path i think it should not result in wrong references resolving)

**Comment by rneatherway** [comment]
I took another look and in fact the error message is quite good here:

> The OutputPath property is not set for project 'CalculatorService.fsproj'.  Please check to make sure that you have specified a valid combination of Configuration and Platform for this project.  Configuration='Debug'  Platform='AnyCPU'. 

The two places where you specify the `OutputPath` are both guarded by platform `x64`, so under the defaults set by the upper section, which sets platform to `AnyCPU` they will not be used. Probably when you open this in Visual Studio the solution context sets the platform to `x64`. If I set that property then FCS gets better options (see bottom of this comment).

> (Also even if something would be wrong with output path i think it should not result in wrong references resolving)

The project cracking functionality works by invoking MSBuild, so it's really up to whatever MSBuild does. It's now a standalone application: https://github.com/fsharp/FSharp.Compiler.Service/blob/master/src/fsharp/FSharp.Compiler.Service.ProjectCrackerTool/Program.fs

If we want to make the errors easier to present, that's another question. However, in the past the build has often failed ([would check here](https://github.com/fsharp/FSharp.Compiler.Service/blob/master/src/fsharp/FSharp.Compiler.Service.ProjectCrackerTool/Program.fs#L152)) while still providing useful references, so I'm not 100% sure what to do. Perhaps if no references are returned? Always using verbose mode is also likely to slow things down, but then we don't often crack projects so maybe that's OK.


```
~/dev/ServiceFabricSuave (master|✔)
$ mono ../FsAutoComplete/packages/FSharp.Compiler.Service.ProjectCracker/lib/net45/FSharp.Compiler.Service.ProjectCrackerTool.exe src/CalculatorService/CalculatorService.fsproj false Platform x64 | jq .
{
  "LogOutput@": "",
  "Options@": [
    "--simpleresolution",
    "--noframework",
    "--out:/Users/robnea/dev/ServiceFabricSuave/src/CalculatorService/bin/Debug/CalculatorService.exe",
    "--doc:/Users/robnea/dev/ServiceFabricSuave/src/CalculatorService/bin/Debug/SuaveHost.XML",
    "--fullpaths",
    "--flaterrors",
    "--target:exe",
    "--define:DEBUG",
    "--define:TRACE",
    "--debug+",
    "--optimize-",
    "--tailcalls-",
    "--debug:full",
    "--platform:x64",
    "-r:/usr/local/Cellar/mono/4.2.1.102_1/lib/mono/4.5/System.Runtime.Serialization.dll",
    "-r:/usr/local/Cellar/mono/4.2.1.102_1/lib/mono/4.5/System.ServiceModel.dll",
    "-r:/usr/local/Cellar/mono/4.2.1.102_1/lib/mono/4.5/mscorlib.dll",
    "-r:/usr/local/Cellar/mono/4.2.1.102_1/lib/mono/4.5/FSharp.Core.dll",
    "-r:/usr/local/Cellar/mono/4.2.1.102_1/lib/mono/4.5/System.dll",
    "-r:/usr/local/Cellar/mono/4.2.1.102_1/lib/mono/4.5/System.Numerics.dll",
    "-r:/usr/local/Cellar/mono/4.2.1.102_1/lib/mono/4.5/System.Core.dll",
    "/Users/robnea/dev/ServiceFabricSuave/src/CalculatorService/AssemblyInfo.fs",
    "/Users/robnea/dev/ServiceFabricSuave/src/CalculatorService/ServicesCommon.fs",
    "/Users/robnea/dev/ServiceFabricSuave/src/CalculatorService/Services.fs",
    "/Users/robnea/dev/ServiceFabricSuave/src/CalculatorService/Program.fs"
  ],
  "ProjectFile@": "src/CalculatorService/CalculatorService.fsproj",
  "ReferencedProjectOptions@": []
}
```

**Comment by davidpodhola** [comment]
Happening to me in both Atom and VS Code too (Windows 10 Enterprise 10.0.11099, Visual Studio 2015 Enterpise 2015 Version 14.0.24720.00 Update 1) in [console](https://github.com/davidpodhola/sagacious-octo-kumquat/tree/master/console) project. Is there any way I can help with this?
![screenshot 1](https://cloud.githubusercontent.com/assets/436605/12374563/5b272e66-bc9f-11e5-8e88-eaea846f6647.png)

**Comment by rneatherway** [comment]
Hi @davidpodhola, thanks a lot for offering to help. Can you please find the file `ProjectCrackerTool.exe` on your computer (will be wherever you install your Atom packages), and run the following from a command prompt:

FSharp.Compiler.Service.ProjectCrackerTool.exe --text <path to>/console.fsproj true

@Krzysztof-Cieslak it would be great if there was a way to toggle this verbose logging in Atom/VSCode.

**Comment by davidpodhola** [comment]
Thanks @rneatherway, this is the result
```
C:\Users\david.podhola\.atom\packages\ionide-fsharp\bin>FSharp.Compiler.Service.ProjectCrackerTool.exe --text C:\Users\david.podhola\S\sagacious-octo-kumquat\console\console.fsproj true
{ProjectFile =
  "C:\Users\david.podhola\S\sagacious-octo-kumquat\console\console.fsproj";
 Options =
  [|"--simpleresolution"; "--noframework";
    "--doc:C:\Users\david.podhola\S\sagacious-octo-kumquat\console\bin\Debug\console.XML";
    "--fullpaths"; "--flaterrors"; "--target:exe"; "--define:DEBUG";
    "--define:TRACE"; "--debug+"; "--optimize-"; "--tailcalls-"; "--debug:full";
    "--platform:anycpu";
    "C:\Users\david.podhola\S\sagacious-octo-kumquat\console\console.fs"|];
 ReferencedProjectOptions = [||];
 LogOutput =
  "Build started.
Project "console.fsproj" (Build target(s)):
Building with tools version "14.0".
Project file contains ToolsVersion="12.0". This toolset may be unknown or missing, in which case you may be able to resolve this by installing the appropriate version of MSBuild, or the build may have been forced to a particular ToolsVersion for policy reasons. Treating the project as if it had ToolsVersion="14.0". For more information, please see http://go.microsoft.com/fwlink/?LinkId=293424.
The target "Build" does not exist in the project.
Done building project "console.fsproj" -- FAILED.
Build FAILED.
";}
```
Might be related to this section in my `fsproj`
```
  <Choose>
    <When Condition="'$(VisualStudioVersion)' == '11.0'">
      <PropertyGroup Condition="Exists('$(MSBuildExtensionsPath32)\..\Microsoft SDKs\F#\3.0\Framework\v4.0\Microsoft.FSharp.Targets')">
        <FSharpTargetsPath>$(MSBuildExtensionsPath32)\..\Microsoft SDKs\F#\3.0\Framework\v4.0\Microsoft.FSharp.Targets</FSharpTargetsPath>
      </PropertyGroup>
    </When>
    <Otherwise>
      <PropertyGroup Condition="Exists('$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)\FSharp\Microsoft.FSharp.Targets')">
        <FSharpTargetsPath>$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)\FSharp\Microsoft.FSharp.Targets</FSharpTargetsPath>
      </PropertyGroup>
    </Otherwise>
  </Choose>
  <Import Project="$(FSharpTargetsPath)" Condition="Exists('$(FSharpTargetsPath)')" />
```
because if I simplify it to just
```
  <Import Project="C:\Program Files (x86)\Microsoft SDKs\F#\4.0\Framework\v4.0\Microsoft.FSharp.targets" />
```
I get the attachment which I assume is better. Not sure how the previous fsproj was working anyway.
[result.txt](https://github.com/ionide/ionide-fsharp/files/93461/result.txt)
**Anyway the result in Atom is still the same (the namespace or module 'PostSharp' is not defined)**

**Comment by rneatherway** [comment]
OK, I think that the first problem you experienced is probably because the project cracker assumes Visual Studio 12 in one part (https://github.com/fsharp/FSharp.Compiler.Service/blob/master/src/fsharp/FSharp.Compiler.Service.ProjectCrackerTool/Program.fs#L131).

Can you tell me the contents of this folder:

    C:\Program Files (x86)\MSBuild\Microsoft\VisualStudio

and which versions of VS you have installed please? If you can clone FCS, change that line to 14.0 and give it another guy against the original project file that would be even better.

In the second version the listed references include PostSharp, so I can't see why there would be a problem with missing reference at that stage.

**Comment by davidpodhola** [comment]
To answer the question: just two folders there, `v12.0` and `v14.0`. I have `Visual Studio 2015 Enterpise 2015 Version 14.0.24720.00 Update 1` installed only, the subfolder `v12.0` containts the folder `TypeScript` only.

**Comment by avy-the-one** [comment]
Here's a very dirty hack i have used to overcome namespaces issue.
 
Works in Windows with Atom and VsCode. Im not using Visual Studio in any way (although, i have it installed), so there is no *.fsproj, just VsCode (or Atom) + Paket + Fake.

1. Use *.fsx files instead of *.fs (rename all your *.fs to *.fsx) to get access to fsi directives (`#load`, `#r`), otherwise it will complain that you can't use directives.
2. In each file you can use `#load` directive to reference file you need types from.
Also, you can wrap it with `#if INTERACTIVE ... #endif`
3. Notice, everything works fine now

Also notice, im using this environment setup in a new project, so i have no legacy code and it's not a problem for me to just drop in *.fsx files instead of *.fs

Please, let me know if it is really a wrong way of doing things :smile:

**Comment by davidpodhola** [comment]
@avy-the-one thanks, I will try, my problem sith `fsx` files is I am not used to use them. I mostly create the projects with `fsproj` files.

**Comment by rneatherway** [comment]
Please don't go down that road! You will have to manually maintain the references in each file. I have an update at the referenced PR, which should address this problem. Once it is merged into FCS then we can do new releases of FSAC and Ionide.

**Comment by rneatherway** [comment]
@Krzysztof-Cieslak I've released a new version of FsAutoComplete with latest FCS that should fix at least some of these issues. Can you update ionide to include it?

**Comment by Krzysztof-Cieslak** [comment]
New FSAC released in `1.6.1`

**Comment by davidpodhola** [comment]
working for me, thanks a lot!

**Comment by rneatherway** [comment]
Great!

**Comment by kunjee17** [comment]
Hi, I am facing similar kind of issue. When I am having multiple projects.

I am having solution having C# and F# projects in src/<projectName> folder. If I open top level folder F# project can't resolve the namespace. But if I go to F# project folder and open the project it works fine. This issue is only reproducible for Windows machine only. In OSX everything works fine. 

Should I open this issue or file another one?
cc/ @Krzysztof-Cieslak
